src/mo_frontend/typing.ml

```diff
@@ -3221,27 +3318,42 @@ and gather_dec env scope dec : Scope.t =
       fld_src_env = scope.fld_src_env;
     }
 
-and gather_pat env ve pat : Scope.val_env =
-   gather_pat_aux env Scope.Declaration ve pat
+and gather_pat env (scope : Scope.t) pat : Scope.t =
+   gather_pat_aux env Scope.Declaration scope pat
 
-and gather_pat_aux env val_kind ve pat : Scope.val_env =
+and gather_pat_aux env val_kind scope pat : Scope.t =
   match pat.it with
-  | WildP | LitP _ | SignP _ -> ve
-  | VarP id -> gather_id env ve id val_kind
-  | TupP pats -> List.fold_left (gather_pat env) ve pats
-  | ObjP pfs -> List.fold_left (gather_pat_field env) ve pfs
+  | WildP | LitP _ | SignP _ -> scope
+  | VarP id -> Scope.adjoin_val_env scope (gather_id env scope.Scope.val_env id val_kind)
+  | TupP pats -> List.fold_left (gather_pat env) scope pats
+  | ObjP pfs -> List.fold_left (gather_pat_field env) scope pfs
   | TagP (_, pat1) | AltP (pat1, _) | OptP pat1
-  | AnnotP (pat1, _) | ParP pat1 -> gather_pat env ve pat1
+  | AnnotP (pat1, _) | ParP pat1 -> gather_pat env scope pat1
 
-and gather_pat_field env ve pf : Scope.val_env =
+and gather_pat_field env scope pf : Scope.t =
   let val_kind = kind_of_field_pattern pf in
-  gather_pat_aux env val_kind ve pf.it.pat
+  match pf.it with
+  | VarPF (id, pat) -> gather_pat_aux env val_kind scope pat
+  | TypPF id -> gather_typ_id env scope id
 
 and gather_id env ve id val_kind : Scope.val_env =
   if T.Env.mem id.it ve then
     error_duplicate env "" id;
   T.Env.add id.it (T.Pre, id.at, val_kind) ve
 
+and gather_typ_id env scope id : Scope.t =
+  let open Scope in
+  if T.Env.mem id.it scope.typ_env then
+    error_duplicate env "type " id;
+  (* NOTE: If we decide to require specifying the arity and bounds of
+  type constructors on type pattern fields we need to record them here *)
+  let pre_k = T.Abs ([], T.Pre) in
```

Comment:
```suggestion
  let pre_k = T.Def ([], T.Pre) in
```
Just for consistency, I would expect these to be Def, not Abs (which we use for type parameters). Don't think it matters though.

I'm a little suprised you can get away with the incorrect arity [] though...

Reply:
(Maybe we should add a placeholder kind PreK for unkown...)